mpitests/test_expval.py

```diff
@@ -22,114 +22,260 @@
 from conftest import PHI, THETA, VARPHI, device_name
 from mpi4py import MPI
 
+numQubits = 8
 
-@pytest.mark.parametrize("theta, phi", list(zip(THETA, PHI)))
-class TestExpval:
-    """Test expectation values"""
 
-    def test_identity_expectation(self, theta, phi, tol):
-        """Test that identity expectation value (i.e. the trace) is 1"""
-        dev = qml.device(device_name, mpi=True, wires=3)
+def create_random_init_state(numWires, C_DTYPE, seed_value=48):
+    """Returns a random initial state of a certain type."""
+    np.random.seed(seed_value)
 
-        O1 = qml.Identity(wires=[0])
-        O2 = qml.Identity(wires=[1])
+    R_DTYPE = np.float64 if C_DTYPE == np.complex128 else np.float32
 
-        dev.apply(
-            [qml.RX(theta, wires=[0]), qml.RX(phi, wires=[1]), qml.CNOT(wires=[0, 1])],
-            rotations=[*O1.diagonalizing_gates(), *O2.diagonalizing_gates()],
-        )
+    num_elements = 2**numWires
+    init_state = np.random.rand(num_elements).astype(R_DTYPE) + 1j * np.random.rand(
+        num_elements
+    ).astype(R_DTYPE)
 
-        res = np.array([dev.expval(O1), dev.expval(O2)])
-        assert np.allclose(res, np.array([1, 1]), tol)
+    init_state = init_state / np.linalg.norm(init_state)
+    return init_state
 
-    def test_pauliz_expectation(self, theta, phi, tol):
-        """Test that PauliZ expectation value is correct"""
-        dev = qml.device(device_name, mpi=True, wires=3)
 
-        O1 = qml.PauliZ(wires=[0])
-        O2 = qml.PauliZ(wires=[1])
+def apply_operation_gates_qnode_param(tol, dev_mpi, operation, par, Wires):
+    """Wrapper applying a parametric gate with QNode function."""
+    num_wires = numQubits
+    comm = MPI.COMM_WORLD
+    commSize = comm.Get_size()
+    num_global_wires = commSize.bit_length() - 1
+    num_local_wires = num_wires - num_global_wires
 
-        dev.apply(
-            [qml.RX(theta, wires=[0]), qml.RX(phi, wires=[1]), qml.CNOT(wires=[0, 1])],
-            rotations=[*O1.diagonalizing_gates(), *O2.diagonalizing_gates()],
-        )
+    c_dtype = dev_mpi.c_dtype
 
-        res = np.array([dev.expval(O1), dev.expval(O2)])
-        assert np.allclose(res, np.array([np.cos(theta), np.cos(theta) * np.cos(phi)]), tol)
+    expected_output_cpu = np.zeros(1 << num_wires).astype(c_dtype)
```

Comment:
can we do it in python style like `2**num_wires`?

Reply:
Done